### ISSUE TITLE
Syntax sugar for accessing props

### ISSUE THREAD

**Issue by nickmccurdy** [planttheidea/selectorator#9]
The String property syntax is very convenient for writing selectors, but it seems to only be supported on state. It would be nice to be able to use this syntax to access props rather than Reselect's convention of `(state, props) => props.something`. If a plain string syntax isn't possible, perhaps we could expose the function that returns a getter from a string to make it available to prop selectors.

**Comment by planttheidea** [comment]
Well this may be easy to implement, but ideally we would do this in a way that doesn't break existing usage but also handles the wacky edge cases. The convention you describe is just that, a convention, because if you look at the `reselect` source code you'll see that you can pass any number of arguments to the selector and it passes those through to your value selectors. As such, theoretically you could use the convenience syntax from the third, fourth, nth parameter ... far less likely, but technically possible. This means allowing selection of any argument index for shorthand purposes.

I can think of two possible ways to accomplish this:

## Add an option

This is easy, just add an option that identifies where shorthand selectors are pulled from:

```javascript
const example = createSelector(['some.item'], (item) => console.log(item), {shorthandArgumentIndex: 1);

const state = {not: 'accessed'};
const props = {some: {item: 'accessed'}};

example(state, props); // 'accessed'
```

Not a lot of drawbacks here to me, but feel free to mention any you see.

## Sub-function

Create an extension of the existing `createSelector` with a sub-function, something like:

```javascript
const example = createSelector.shorthandArg(2)(['some.item'], (item) => console.log(item));

const state = {not: 'accessed'};
const props = {some: {item: 'accessed'}};

example(state, props); // 'accessed'
```

Which would be the "solves all" usage, and then I could create a simple convenience function for the common convention use-case you describe:

```javascript
const example = createSelector.fromProps(['some.item'], (item) => console.log('item'));

const state = {not: 'accessed'};
const props = {some: {item: 'accessed'}};

example(state, props); // 'accessed'
```

The only weirdness here is when you compose this with other selectors in the list, because that argument index seletction would only apply to the shorthand declarations, regular selectors would still get all the parameters.

Thoughts?

**Comment by nickmccurdy** [comment]
It seems like we just need a way to sort each selector argument into an argument index. I prefer the sub-function over the object, but I think it would be even better to support a 2D array instead. This won't require any new functions but could remain back compatible (if the array is only 1D you just default to the first argument). Then we flatten the arguments into a single array of arguments in order, using the appropriate argument index for each inner array. Example: `createSelector([['state'], ['prop']], (state, prop) =>

**Comment by planttheidea** [comment]
I mean I see what you're doing here, but this creates a highly implicit API, and I don't like that cognitive load. How about using objects instead?

```javascript
const example = createSelector([{key: 'some.item', argIndex: 1}, 'item'], (propsItem, stateItem) => console.log(`${propsItem} ${stateItem}`));

const state = {item: 'from state'};
const props = {some: {item: 'from props'}};

example(state, props); // from props from state
```
Easy to implement in code, easy to understand what is going on, allows combination of both use-cases.

**Comment by nickmccurdy** [comment]
Nice, I like that syntax more. I think it's a good compromise.

**Comment by planttheidea** [comment]
Cool, I'll work on this and hopefully have a PR up shortly.

**Comment by planttheidea** [comment]
Alright PR created ... forgive the bloat, I haven't touched this repo in a long time and I couldn't help myself with cleanup of old code. I left a comment on the meaty change for this to happen.

**Comment by planttheidea** [comment]
Changes released under `3.3.0` / `1.4.0`. Thanks for the idea!
